Update SSML package with parameter validation, documentation, & examples

harrisonhjones opened the pull request:
# Overview

- Added `Deprecated` notice to `SSMLTextBuilder` type & functions in `skillserver` package
- Added flexible parameters & parameter validation to a few `Append` functions
- Added documentation
- Added examples

# Details

- Added `Deprecated` notice to SSML builder type & funcs in skillserver package
- Added flexible parameters & parameter validation to `AppendBreak`
 - Now takes either a `pause.Strength` param or a `time.Duration`
- Added parameter validation to `AppendAudio`
 - Now checks to make sure it's a valid HTTPS url
- Added flexible parameters & parameter validation to `AppendProsody`
 - Now takes either `prosody.*` constants or `int`s
 - Also makes all parameters optional (with `nil`)
- Added documentation to the package and all functions
 - Exported the `Builder` type so all `Append` functions appear in the generated documentation
- Added examples for all functions
- Append functions now only return an error. The builder doesn't need to be returned since a pointer receiver is being used

# Testing

- 100% unit tests of `ssml` package pass
- 100% example tests of `ssml` package pass

# TODO

- Support for more SSML features

harrisonhjones committed:
Mark old SSML types/funcs as deprecated

harrisonhjones committed:
Export Builder type so methods appear in GoDoc

harrisonhjones committed:
Update AppendBreak to accept either pause strength OR time.Duration

harrisonhjones committed:
Add basic GoDoc to package

harrisonhjones committed:
Add src URL validation to AppendAudio

harrisonhjones committed:
Add flexible parameters & validation to AppendAppendProsody

harrisonhjones committed:
Remove returned Builder from all Append funcs

harrisonhjones committed:
Ensure all doc comments end in a period

harrisonhjones committed:
Add examples

harrisonhjones commented:
@mikeflynn  this PR is ready to review but after thinking about it for a while I wanted your input on my changes to the builder pattern. In the last PR I added the standard go `return error` pattern to each `Append*` function. While go-like this breaks builder chaining `builder.appendX.appendY.build()`. In this PR I completely remove the returned builder because it's largely worthless if it's not the only return value. However, after thinking about it I have a few ideas on how we could return it to using the builder pattern:

1. suppress the error and simply not add text to the builder if a call fails. Subsequent successful calls would still add text to the builder. 
1. panic on error
1. Have the `Build()` function return both a `string` and a slice of `error`s. When a chained call ran into an error it would append that error to the builder internal error slice. `Build()` would return this slice. If there are no errors return a nil slice.

My preference is option 3 as I think it's most go-like. You could still use the standard `if err != nil` check after `Build()`.

rking788 commented:
My vote would be either 1 or 3. I think there are only a few cases where the append methods can actually return a valuable error and in those cases there could be more responsibility put on the user to do input validation to avoid those errors. 

- AppendAudio method could take a URL parameter since the user should be doing some kind of validation that they are passing a valid URL anyway.

- AppendBreak could be split into two overloads, one taking a pause.Strength and one taking a time.Duration to get compile time checks for the user and they use the same internal appendBreak.

- AppendProsody is the tricky one. I think since the expected inputs are well defined i think it would be fine to not append anything and say it is just user error. a more user friendly way of working around it would probably to do something like option 3. I think this is similar to how Go handles SQL transactions. My only recommendation there would be to combine the errors somehow into a single error string so the user can just check `err != nil` instead of `errs != nil && len(errs) > 0`

mikeflynn commented:
I agree that option #3 is the most Go-like and that's the one I would vote for.

I also wanted to discuss how we wanted to handle breaking changes to the skillserver package. I'm wondering if we don't simply freeze skillserver and then once all the other packages are broken out we can make a "skillserver-v2" package that brings them all together. Because of Go's lack of versions, I'm concerned with pushing any breaking changes to skillserver and while skillserver could be updated to use some of the new changes, we should avoid any breaking changes to not impact current projects.

mikeflynn commented:
@harrisonhjones Let me know when this is ready to merge. I'm not sure if you have more to do after your question and if what you thought about my plan to freeze skillserver (or at least no breaking changes).

cc @rking788

rking788 commented:
I agree since this is such a breaking change and the trouble with Go package versioning, it probably makes sense to make a non-backward compatible new package name like you recommended.

harrisonhjones commented:
Hey gang, sorry for the delay on commenting. Work's been busy. Comments below:

> My vote would be either 1 or 3.  / I agree that option #3 is the most Go-like and that's the one I would vote for.

Let's go with option 3 then. I'll update the PR.

> AppendAudio method could take a URL parameter since the user should be doing some kind of validation that they are passing a valid URL anyway.

Not a bad idea and I'm willing to change to this but I think keeping it as a string is more flexible as Amazon might allow for other non-url audio sources in the future.

> AppendBreak could be split into two overloads, one taking a pause.Strength and one taking a time.Duration to get compile time checks for the user and they use the same internal appendBreak.

I'm unsure how to do overloads in Go the way you are describing. Could you provide an example?

> I also wanted to discuss how we wanted to handle breaking changes to the skillserver package. I'm wondering if we don't simply freeze skillserver and then once all the other packages are broken out we can make a "skillserver-v2" package that brings them all together. Because of Go's lack of versions, I'm concerned with pushing any breaking changes to skillserver and while skillserver could be updated to use some of the new changes, we should avoid any breaking changes to not impact current projects.

I'm ok with freezing skill-server but I do think we should make deprecated methods as deprecated once they've been ported to the new packages. With a bit of care I don't think we'll make any breaking changes to skillserver but perhaps I'm missing something. Can you elaborate?

mikeflynn commented:
We should definitely mark things as deprecated as a sign that things won't be getting updated and to alert users to the new broken out packages. What I don't want to have happen is swapping of params or return vars in existing functions, so that when someone sets their existing go-alexa app up on a new machine their app won't start spitting error messages at them.

If we can still upgrade the skillserver package to use the new sister packages without any changing of the existing skillserver API, then great! That's the best situation, but I'd qualify my default position as an "abundance of caution" as the library does have a fair bit of use out in the community.

harrisonhjones committed:
Bring back the builder pattern

- Each builder append function is returned the builder
- Stores errors in a concurrently-safe slice

harrisonhjones committed:
Update examples given new build return values

harrisonhjones committed:
Update tests to deeply check errors

harrisonhjones committed:
Add test for concurrent safety

harrisonhjones commented:
@mikeflynn (and others) the PR has been updated; play take a look when you have time.

As for updating skillserver I took another look and I don't see a way to retrofit it with the `ssml` package without some crappy code so I'm ok to "freeze" skill server (or at least the `ssml-builder` file)

harrisonhjones commented:
A friendly ping to take a look at the PR @mikeflynn . If you are strapped for time (or just want some help) I'd be happy to have a quick discussion about becoming a co-maintainer

mikeflynn commented:
Thanks for the ping. It's been a crazy week! I'll look at this change now.

mikeflynn commented:
Change looks good. Merging in!
